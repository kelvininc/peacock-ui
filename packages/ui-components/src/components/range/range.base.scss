@import '../../assets/styles/globals';

:host {
	/**
	 * @prop --range-width: Width of the range slider
	 * @prop --range-height: Height of the range slider
	 * @prop --range-selector-radius: Radius of the range thumb
	 * @prop --slider-background-filled: color of the slider when its full
	 * @prop --slider-background-empty: color of the slider when its empty
	 * @prop --thumb-border-color: thumb border color
	 * @prop --range-label-color: range labels color
	 * @prop --select-label-color: select labels color
	*/

	--range-width: 100%;
	--range-height: 12px;
	--range-selector-radius: 20px;
}

.range-container {
	width: var(--range-width);
	margin: 24% auto;
	position: relative;
}

.range-min-max {
	@include kv-font-label-small-regular;
	width: 100%;
	display: flex;
	justify-content: space-between;
	margin-top: $spacing-4x;
	user-select: none;
	color: var(--range-label-color);
}

.slider {
	width: var(--range-width);
	height: var(--range-height);
	outline: none;
	margin: 0;
	border-radius: $spacing;
	user-select: none;
	background: var(--slider-background-empty);

	&::-webkit-slider-thumb {
		border: calc($spacing - 1px) solid var(--thumb-border-color);
		background: var(--slider-background-filled);
		height: calc(var(--range-selector-radius) + 5px);
		width: calc(var(--range-selector-radius) + 5px);
		border-radius: 50%;
		cursor: pointer;
		appearance: none;
	}

	&::-moz-range-thumb {
		border: $spacing solid var(--thumb-border-color);
		background: var(--slider-background-filled);
		height: var(--range-selector-radius);
		width: var(--range-selector-radius);
		border-radius: 50%;
		cursor: pointer;
		appearance: none;
	}
}

.select-value {
	@include kv-font-h1-semibold;
	height: 48px;
	top: -68px;
	transform: translateX(-50%);
	position: absolute;
	background-color: transparent;
	text-align: center;
	color: var(--select-label-color);
}

input {
	appearance: none;
}
