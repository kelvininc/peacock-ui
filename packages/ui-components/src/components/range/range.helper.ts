export const getInputPercentageFromValue = (inputValue: string, min: number, max: number) => {
	return (100 * (parseInt(inputValue) - min)) / (max - min);
};

export const getOffset = (percentage: number) => {
	const offSet = -0.25;
	const halfDistance = percentage - 50;

	return halfDistance * offSet;
};
