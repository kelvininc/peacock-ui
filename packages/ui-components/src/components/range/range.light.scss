@import '../../assets/styles/globals';
@import './range.base';

:host {
	--slider-background-filled: #{kv-color('primary')};
	--slider-background-empty: #{kv-color('neutral-3')};
	--thumb-border-color: #{kv-color('primary')};
	--range-label-color: #{kv-color('neutral-5')};
	--select-label-color: #{kv-color('neutral-5')};
}
