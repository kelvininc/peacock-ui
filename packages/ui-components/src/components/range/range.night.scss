@import '../../assets/styles/globals';
@import './range.base';

:host {
	--slider-background-filled: #{kv-color('primary')};
	--slider-background-empty: #{kv-color('neutral-9')};
	--thumb-border-color: #{kv-color('neutral-0')};
	--range-label-color: #{kv-color('neutral-5')};
	--select-label-color: #{kv-color('neutral-2')};
}
