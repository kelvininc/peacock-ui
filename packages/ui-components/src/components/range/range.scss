@import '../../assets/styles/globals';

:host {
	/**
	 * @prop --range-width: Width of the range slider
	 * @prop --slider-background-filled: color of the slider when its full
	 * @prop --slider-background-empty: color of the slider when its empty
	*/

	--range-width: 100%;
	--range-height: 12px;
	--button-selector-width: 20px;
	--button-selector-height: 20px;
	--slider-background-filled: #{kv-color('primary')};
	--slider-background-empty: #{kv-color('neutral-9')};
}

.range-container {
	width: var(--range-width);
	margin: 24% auto;
	position: relative;
}

.range-min-max {
	@include kv-font-label-small-regular;
	width: 100%;
	display: flex;
	justify-content: space-between;
	color: kv-color('neutral-5');
	margin-top: $spacing-4x;
	user-select: none;
}

.slider {
	width: var(--range-width);
	height: var(--range-height);
	outline: none;
	margin: 0;
	border-radius: $spacing;
	background: var(--slider-background-empty);
	user-select: none;

	&::-webkit-slider-thumb {
		border: calc($spacing - 1px) solid kv-color('neutral-0');
		height: calc(var(--button-selector-height) + 5px);
		width: calc(var(--button-selector-width) + 5px);
		border-radius: 50%;
		background: var(--slider-background-filled);
		cursor: pointer;
		appearance: none;
	}

	&::-moz-range-thumb {
		border: $spacing solid kv-color('neutral-0');
		height: var(--button-selector-height);
		width: var(--button-selector-width);
		border-radius: 50%;
		background: var(--slider-background-filled);
		cursor: pointer;
		appearance: none;
	}
}

.select-value {
	@include kv-font-h1-semibold;
	width: 48px;
	height: 48px;
	top: -68px;
	transform: translateX(-50%);
	position: absolute;
	background-color: transparent;
	text-align: center;
	color: kv-color('neutral-2');
}

input {
	appearance: none;
}
