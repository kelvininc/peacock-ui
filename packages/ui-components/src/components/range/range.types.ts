import { EventEmitter } from '@stencil/core';

export interface IRangeEvents {
	/** Emitted when the value changes */
	valueChange: EventEmitter<number>;
}

export interface IRange {
	/** (required) Range minimum value */
	min: number;
	/** (required) Range maximum value */
	max: number;
	/** (optional) Range value */
	value?: number;
	/** (optional) Range value step */
	step?: number;
	/** (optional) Hide value label */
	hideLabel?: boolean;
}
