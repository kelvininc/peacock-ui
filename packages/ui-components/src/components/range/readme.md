# _kv-range_

<!-- Auto Generated Below -->


## Usage

### Angular

<!-- Default -->
<kv-range min=0 max=100 value=0 step=1></kv-range>


### React

import React from 'react';
import { KvRange } from '@kelvininc/react-ui-components';

export const KvRangeExample: React.FC = () => (
  <>
	<KvRange min={0} max={100} step={1} value={0} />
  </>
);



## Properties

| Property         | Attribute    | Description                      | Type                        | Default     |
| ---------------- | ------------ | -------------------------------- | --------------------------- | ----------- |
| `disabled`       | `disabled`   | (optional) Range input disabled  | `boolean`                   | `false`     |
| `hideLabel`      | `hide-label` | (optional) Hide value label      | `boolean`                   | `false`     |
| `max`            | `max`        | (required) Range maximum value   | `number`                    | `undefined` |
| `maxLabel`       | `max-label`  | (optional) Max label             | `string`                    | `undefined` |
| `min`            | `min`        | (required) Range minimum value   | `number`                    | `undefined` |
| `minLabel`       | `min-label`  | (optional) Min label             | `string`                    | `undefined` |
| `step`           | `step`       | (optional) Range value step      | `number`                    | `1`         |
| `value`          | `value`      | (optional) Range value           | `number`                    | `0`         |
| `valueFormatter` | --           | (optional) Range value formatter | `(value: number) => string` | `identity`  |


## Events

| Event         | Description                    | Type                  |
| ------------- | ------------------------------ | --------------------- |
| `valueChange` | Emitted when the value changes | `CustomEvent<number>` |


## CSS Custom Properties

| Name                         | Description                        |
| ---------------------------- | ---------------------------------- |
| `--range-height`             | Height of the range slider         |
| `--range-label-color`        | range labels color                 |
| `--range-selector-radius`    | Radius of the range thumb          |
| `--range-width`              | Width of the range slider          |
| `--select-label-color`       | select labels color                |
| `--slider-background-empty`  | color of the slider when its empty |
| `--slider-background-filled` | color of the slider when its full  |
| `--thumb-border-color`       | thumb border color                 |


----------------------------------------------


