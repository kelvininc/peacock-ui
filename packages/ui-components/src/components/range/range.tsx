import { Component, Host, h, Prop, Element, EventEmitter, Event } from '@stencil/core';
import { IRange, IRangeEvents } from './range.types';
import { getInputPercentageFromValue, getOffset } from './range.helper';

@Component({
	tag: 'kv-range',
	styleUrl: 'range.scss',
	shadow: true
})
export class KvRange implements IRange, IRangeEvents {
	@Element() el!: HTMLKvRangeElement;
	/** @inheritdoc */
	@Prop({ reflect: true }) min: number;
	/** @inheritdoc */
	@Prop({ reflect: true }) max: number;
	/** @inheritdoc */
	@Prop({ reflect: true, mutable: true }) value?: number = 0;
	/** @inheritdoc */
	@Prop({ reflect: true }) step?: number = 1;
	/** @inheritdoc */
	@Event() valueChange: EventEmitter<number>;

	componentDidRender() {
		this.applyCssStyles();
	}

	private applyCssStyles = () => {
		const rangeInputValue = this.el.shadowRoot.querySelector('input');

		const selector = this.el.shadowRoot.getElementById('select-value');

		let inputValue = rangeInputValue.value;
		const percentage = getInputPercentageFromValue(inputValue, this.min, this.max);
		const offSet = getOffset(percentage);

		selector.style.left = percentage + '%';
		selector.style.marginLeft = offSet + 'px';
		rangeInputValue.style.background = `linear-gradient(90deg, var(--slider-background-filled) ${percentage + offSet}%, var(--slider-background-empty) ${percentage}%)`;
		this.onValueChange(parseInt(inputValue));
	};

	private onValueChange = (newValue: number) => {
		this.value = newValue;
		this.valueChange.emit(newValue);
	};

	render() {
		return (
			<Host>
				<div class="range-container">
					<input id="slider" class="slider" type="range" min={this.min} max={this.max} value={this.value} step={this.step} onInput={this.applyCssStyles} />
					<span id="select-value" class="select-value">
						{this.value}
					</span>
					<div class="range-min-max">
						<span>{this.min}</span>
						<span>{this.max}</span>
					</div>
				</div>
			</Host>
		);
	}
}
