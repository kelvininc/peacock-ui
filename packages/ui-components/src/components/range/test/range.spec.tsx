import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { KvRange } from '../range';

describe('Range (unit tests)', () => {
	let page: SpecPage;
	let component: KvRange;

	describe('when uses default props', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvRange],
				html: `<kv-range></kv-range>`
			});
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when user defines props', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvRange],
				html: `<kv-range min=0 max=100 step=10 value=0></kv-range>`
			});
			component = page.rootInstance;
		});

		it('component min value should match the given value', () => {
			expect(component.min).toBe(0);
		});

		it('component max value should match the given value', () => {
			expect(component.max).toBe(100);
		});

		it('component step value should match the given value', () => {
			expect(component.step).toBe(10);
		});

		it('component value prop value should match the given value', () => {
			expect(component.value).toBe(0);
		});
	});

	describe('when user defines only min and max', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvRange],
				html: `<kv-range min=0 max=100></kv-range>`
			});
			component = page.rootInstance;
		});

		it('component min value should match the given value', () => {
			expect(component.min).toBe(0);
		});

		it('component max value should match the given value', () => {
			expect(component.max).toBe(100);
		});

		it('component step value should match the given value', () => {
			expect(component.step).toBe(1);
		});

		it('component value prop value should match the given value', () => {
			expect(component.value).toBe(0);
		});
	});
});
