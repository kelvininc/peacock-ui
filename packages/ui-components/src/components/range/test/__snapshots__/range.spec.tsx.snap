// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Range (unit tests) when uses default props should match the snapshot 1`] = `
<kv-range step="1" value="0">
  <mock:shadow-root>
    <div class="range-container">
      <input class="slider" id="slider" step="1" type="range" value="0" style="background: linear-gradient(90deg, var(--slider-background-filled) NaN%, var(--slider-background-empty) NaN%);">
      <span class="select-value" id="select-value" style="left: NaN%; margin-left: NaNpx;">
        0
      </span>
      <div class="range-min-max">
        <span></span>
        <span></span>
      </div>
    </div>
  </mock:shadow-root>
</kv-range>
`;
